### DISCUSSION TITLE
Using int to define midi channel

### DISCUSSION THREAD

**Discussion by ACuppaJoes** [tttapa/Control-Surface#551] [General]
Hey Pieter, 
I would like to be able to change a note's midi channel on the fly. Is this possible and how can I achieve this?

```
int selMIDIch=1;

NoteButton button {
  5,                       // Push button on pin 5
  {MIDI_Notes::C(4), selMIDIch} // Note C4 on MIDI channel 1
  };
```
something along the lines of the code above

**Comment by tttapa** [comment]
Please see https://tttapa.github.io/Control-Surface-doc/Doxygen/da/dc1/FAQ.html#faq-change-address-runtime.

**Reply by ACuppaJoes** [reply]
I am making a midi controller with a menu system that allows the user to change the outputted midi notes using presets. I would also like to give the user to option to globally select midi channels 1-16 in the settings area of the menu. The menu system works by calling a user-defined function when a specific preset is chosen. Using the method shown in the FAQ, does that mean my only option is to duplicate my all my preset functions 16 times and then have the menu system call them individually ? This seems inefficient. Is there no way I can use an int and simply change the int value using the menu?

**Reply by tttapa** [reply]
The problem with changing addresses/channels on the fly, and the reason that banks were introduced in the first place is that you have to save the current address whenever you press a button.

Consider:
- Button pressed: MIDI Note On C4 Ch.1 is sent
- Channel is changed to Ch.2.
- Button released: MIDI Note Off C4 Ch.2 is sent

This is wrong, C4 on Ch.1 will keep playing indefinitely.
The correct behavior is:
- Button pressed: save address C4 Ch.1, then send MIDI Note On for this address
- Channel is changed to Ch.2.
- Button released: send MIDI Note Off for saved address C4 Ch.1, then clear saved address

This saving of addresses (bank settings) is handled by the `Bankable` mechanism of the library. This works fine for applications where you actually want to layer your controller in banks, but if you want full control and set arbitrary addresses for each button, it might not be the best solution.

In this case, you might want to do something like this: 
- Create a class that allows you to select any arbitrary address.
- Add a `getActiveAddress()` method that returns the address that should be sent over MIDI.
- Add `lock()` and `unlock()` methods that save the current address and remove the saved address respectively.
- Compose this class with the `Bankable::MIDI*` classes.

For example:
```cpp
#include <Control_Surface.h>

uint8_t global_channel_setting = 0; // channels are zero-based

class ArbitraryAddress {
  public:
    ArbitraryAddress(uint8_t address) : address(address) {}

    MIDIAddress getCurrentActiveAddress() const {
      return { address, Channel(global_channel_setting) };
    }
    MIDIAddress getActiveAddress() const {
      return saved_address ? saved_address : getCurrentActiveAddress();
    }
    void lock() { saved_address = getCurrentActiveAddress(); }
    void unlock() { saved_address = MIDIAddress::invalid(); }
    
    uint8_t address;

  private:
    MIDIAddress saved_address = MIDIAddress::invalid();
};

// NoteButton alternative that allows you to select any address.
// This is a simple wrapper, it doesn't have any logic of its own, it uses the existing
// support for NoteButton and tells it to use the ArbitraryAddress class when sending MIDI
struct AANoteButton : Bankable::MIDIButton<ArbitraryAddress, DigitalNoteSender> {
  AANoteButton(pin_t pin, uint8_t address, uint8_t velocity = 0x7F)
    : MIDIButton{address, pin, velocity} {}
  void setAddress(uint8_t addr) { address.address = addr; }
  uint8_t getAddress() const { return address.address; }
};

// -----

USBDebugMIDI_Interface midi; // for testing

AANoteButton notebuttons[2] {
  {2, MIDI_Notes::C(4)}, // pin, initial address
  {3, MIDI_Notes::D(4)},
};

Button channel_btn = 5;
Button address_btn = 7;

void setup() {
  Control_Surface.begin();
  channel_btn.begin();
  address_btn.begin();
}

void loop() {
  Control_Surface.loop();

  // Increment the global channel setting if button is pressed
  if (channel_btn.update() == Button::Falling) {
    ++global_channel_setting;
    if (global_channel_setting == 16)
      global_channel_setting = 0;
    Serial << Channel(global_channel_setting) << endl;
  }

  // Change the address of the first notebutton if button is pressed
  if (address_btn.update() == Button::Falling) {
    if (notebuttons[0].getAddress() == MIDI_Notes::C(4))
      notebuttons[0].setAddress(MIDI_Notes::C(5));
    else
      notebuttons[0].setAddress(MIDI_Notes::C(4));
  }
}
```

I'd suggest moving everything above the `// -----` into a separate file.

**Edit**: Forgot to mention: Previously, the `address` field was `private`, I just made them `protected` to allow for this use case, so you'll have to update to at least commit 8da3d0c71186ee8b538fa9a8ab9140bcf8cd5309.

**Comment by ACuppaJoes** [comment]
Excuse my lack of knowledge, but how can I now use this to modify an integer in order to set the midi channel address?

In my menu code, this selection field allows me to select and change an integer value. How can I direct this to change the global channel setting? 
```
int selMIDIch=0;
SELECT(selMIDIch,selCh,"Midi Ch",doNothing,noEvent,wrapStyle
 // (name, variable, call to function, occurrence of function)
  ,VALUE("1",0,doNothing,noEvent)
  ,VALUE("2",1,doNothing,noEvent)
  ,VALUE("3",2,doNothing,noEvent)
  ,VALUE("4",3,doNothing,noEvent)
  ,VALUE("5",4,doNothing,noEvent)
  ,VALUE("6",5,doNothing,noEvent)
  ,VALUE("7",6,doNothing,noEvent)
  ,VALUE("8",7,doNothing,noEvent)
  ,VALUE("9",8,doNothing,noEvent)
  ,VALUE("10",9,doNothing,noEvent)
  ,VALUE("11",10,doNothing,noEvent)
  ,VALUE("12",11,doNothing,noEvent)
  ,VALUE("13",12,doNothing,noEvent)
  ,VALUE("14",13,doNothing,noEvent)
  ,VALUE("15",14,doNothing,noEvent)
  ,VALUE("16",15,doNothing,noEvent)
```

Correct me if I am wrong, but I assume I just need to integrate the code above "// NoteButton alternative that allows you to select any address." and change my int name from selMIDIch to global_channel_setting?


Thanks

**Reply by tttapa** [reply]
> Correct me if I am wrong, but I assume I just need to integrate the code above "// NoteButton alternative that allows you to select any address." and change my int name from selMIDIch to global_channel_setting?

Without knowledge of the library you're using, yes, it looks like that should work.

**Reply by ACuppaJoes** [reply]
I am having a small issue. I ran the code below to test and the channel is still not changing. When I check my midi monitor, the midi note is still on channel 1.

```
#include <Control_Surface.h>

uint8_t global_channel_setting = 4; // channels are zero-based

class ArbitraryAddress {
  public:
    ArbitraryAddress(uint8_t address) : address(address) {}

    MIDIAddress getCurrentActiveAddress() const {
      return { address, Channel(global_channel_setting) };
    }
    MIDIAddress getActiveAddress() const {
      return saved_address ? saved_address : getCurrentActiveAddress();
    }
    void lock() { saved_address = getCurrentActiveAddress(); }
    void unlock() { saved_address = MIDIAddress::invalid(); }
    
    uint8_t address;

  private:
    MIDIAddress saved_address = MIDIAddress::invalid();
};

//-----------NOTE BUTTON---------//

NoteButton button {
  5,                       // Push button on pin 5
  {MIDI_Notes::C(4)}



void setup() {
  Control_Surface.begin();

}
void loop() {
  Control_Surface.loop(); // Refresh all elements
  
}

```

**Reply by tttapa** [reply]
You're using a normal `NoteButton` which knows nothing of the `ArbitraryAddress` class. You have to use `AANoteButton`.

**Reply by ACuppaJoes** [reply]
Yes that was the issue, also I forgot to include the struct in my code

**Comment by starwalker00** [comment]
If we discard the "On/Off channels" for which issues can understandably arise easily.
Something like the following should work for use with "normal" channels like Pan/Volume/Expression etc.. right ?

``` c
/**
* reads midi channel on A0
* reads midi value on A1
*/

#include <Control_Surface.h>

USBMIDI_Interface midi;
FilteredAnalog<> analog = A1;

void setup() {
  Serial.begin(9600);
  FilteredAnalog<>::setupADC();
}

void loop() {
  int sensorValue = analogRead(A0);
  midiChannel = map(sensorValue, 0, 1023, 0, 127);

  static Timer<millis> timer = 1; // ms
  if (timer && analog.update())
    Serial.println(analog.getValue());
    midi.sendCC(midiChannel,analog.getValue());
}
```

right ?
